Report section numbers in MS word.

feerrenrut opened the pull request:
Fix for #5946 and #6632

When reading through a document, report when the section number changes. This is tied to the report page number setting (available in document formatting settings dialog).

When using the read paragraph command the "page break" is read when encountering a new section, however when reading by line it is not. This is a pre-existing issue, perhaps this should be read as "break" rather than "page break"

"Continuous section breaks" can be identified by the section number being reported, but no page number. When a "next page section break" is encountered, both the section number and the page number are reported.

The approach taken is to use the range information method to query the current section number. An alternative approach of caching the section in the whole document was attempted however after timing the two approaches it was found that the single call to the range information (at ~ 3ms) was faster than the 3 + 4n (where n is the number of sections in the document) calls required to cache the section information (at ~ 17ms). Interestingly in the cached approach it was found that the size of the document had little effect on the time taken to expand to "story" (the whole document), and increasing the number of sections had quite a small effect on the overall time. 

I have not found a way to get the information about columns yet. I investigated the following range information enums to get the column number, however:
- wdEndOfRangeColumnNumber (17) and wdStartOfRangeColumnNumber (16) both relate to tables only
- wdFirstCharacterColumnNumber (9) relates to the number of letters in the line.

feerrenrut committed:
Report section numbers in MS word.

Fixes #5946. When reading through a document, report when the section
number changes. This is tied to the report page number setting
(available in document formatting settings dialog)

feerrenrut commented:
@michaelDCurran Could you take a look at this?

michaelDCurran commented:
This looks good so far.
It may be possible to detect the type of break. page / section breaks in Microsoft word are represented by a 0x0c character. If the section number changes directly after the break, then it it is a section break. We can alsoo tell what kind of section break it is by getting the section object and looking at the sectionStart property.
Column breaks are represented in Microsoft Word with a 0x0e character. We currently say nothing for these.
We can detect the number of columns for a section using the section's pageSetup's textColumns.count property. It may be possible to look at the range's current offset in points from the left edge of the page (if that can be fetched) and then calculate which column it is in, as each column  exposes its width via TextColumns.

feerrenrut commented:
Thanks Mick. Would you like me to try to address these issues in this PR?

Note to self:

Example of getting the `sectionStart` property, (from word 2003) docs:
https://msdn.microsoft.com/en-us/library/office/aa196636(v=office.11).aspx

`PageSetup.SectionStart` docs:
`PageSetup` can be accessed via the `section` object.
https://msdn.microsoft.com/en-us/library/office/ff193079.aspx

Docs for the `wdSectionStart` enum:
This includes a value for columns too.
https://msdn.microsoft.com/en-us/library/office/ff840975.aspx

Docs for the `break` object:
https://msdn.microsoft.com/en-us/library/office/ff821134.aspx

feerrenrut committed:
Report section and column breaks

The type of break ( 0x0c char ) is now announced, this may be a continuous
section break, a new column section break, new page section break, even
pages section break, odd pages section break.

Column breaks ( 0x0e char) are also announced.

feerrenrut commented:
We can currently report when the column changes and I have been looking into reporting the column number the range is currently in. Here is the approach I have at the moment. I am a little concerned this will be fragile, particularly in regards to right to left setups.

Getting the position of the range. The only way I found to do this was with the [`Window.GetPoint` method](https://msdn.microsoft.com/en-us/library/office/ff836626.aspx). This returns the values in screen coordinates for the arguments: `left`, `top`, `width`, `height`. There is an `enum`, `WdRelativeHorizontalPosition` I'm curious about, particularly its `wdRelativeHorizontalPositionPage` value. I could not find how to use this to "get" information.

Assuming we have the start position of the range, we then need the start position of all the columns on the page in order to know which one the range is in. Luckily `range.pageSetup.textColumns` has all the columns even when the range does not span multiple columns. Using `range.pageSetup.textColumns.item(n).width` gives the width of the column in `points`. The only other space related property on a `TextColumn` object is `TextColumn.SpaceAfter` which returns or sets the amount of spacing (in points) after the specified paragraph or text column. I'm not sure if this is vertical or horizontal space when used with a `TextColumn`.

Since the units for these (`Window.GetPoint` and `TextColumn.Width`) are different (screen coordinates vs points), we need a conversion function:


### Screen To Client

```
BOOL ScreenToClient(
  _In_ HWND    hWnd,
       LPPOINT lpPoint
);

- winUser.h
- ScreenToClient function converts the screen coordinates of a specified point on the screen to client-area coordinates.
- The new coordinates are relative to the upper-left corner of the specified window's client area.
- The ScreenToClient function assumes the specified point is in screen coordinates.
- All coordinates are in device units.
- Do not use ScreenToClient when in a mirroring situation, that is, when changing from left-to-right layout to right-to-left layout. Instead, use MapWindowPoints.
```

### Map Window Points

```
int MapWindowPoints(
  _In_    HWND    hWndFrom,
  _In_    HWND    hWndTo,
  _Inout_ LPPOINT lpPoints,
  _In_    UINT    cPoints
);

- winUser.h
```

### Application.PixelsToPoints  Method
Converts a measurement from points to pixels. Returns the converted measurement as a Single.

```
PixelsToPoints (Points, fVertical)
- Pixels - Required - Type: Single - The pixel value to be converted to points.
- fVertical - Optional - Type: Variant - True to convert vertical pixels; False to convert horizontal pixels.
```

`PixelsToPoints` sounds like it is the most useful, mostly because we do not have a known start position for the `textColumn`. Using points we may be able to work it out by adding margins etc.

## Some real values:

When sending the `getPoint` and the column width result to the log, where `left` is the left coord for the range, as returned by `GetPoint(out left, out top, out width, out height, Range r)` and `ItemNumber` is the `textColumn` item number:

```
ItemNumber: 1 Left: 2660 columnWidth: 207.95
ItemNumber: 2 Left: 2660 columnWidth: 207.95
```

Then trying to find a way to add it up:

```
>>> t._rangeObj.application.windows.count
1
>>> t._rangeObj.application.windows[1].width #  width of the specified document window (in points)
1406
>>> t._rangeObj.application.pointsToPixels(207.95, False) # textColumn width in points
277.0
>>> t._rangeObj.application.pixelsToPoints(2660.0, False) # range "left" position in screen coords
1995.0
>>> t._rangeObj.pageSetup.pageWidth # in points
595.2999877929688
>>> 207.95*2 # because there are two textColumns at 207.95 each
415.9
>>> t._rangeObj.pageSetup.leftMargin
72.0
>>> t._rangeObj.pageSetup.rightMargin
72.0
>>> 73*2 + 207.95*2 # two margins + two text columns
561.9
t._rangeObj.pageSetup.textColumns[1].spaceAfter
35.400001525878906
>>> 595.3 - (72.0 + 207.95 + 35.4 + 207.95 +72.0) # pageWidth - ( left margin + column + spaceAfter + column + right margin )
0.0
```

feerrenrut commented:
Couple of extra notes:

- We need to take into account "gutters". Are there any more concepts that can offset the left or right?
- We need to get the distance from the window border to the start of the "page". Is there another method I've missed that lets you get the start of the range relative to the document? Or perhaps a way to get the screen space of the column? Can we get the range of the column?

Ideas that don't work:
- Get the whole document, look for the column break `0x0e` characters, use GetPoint to get the position of the range just after the character.
 - There is no need for a column break character, a document can have multiple columns with no breaks.
- Use Application.Windows(n).Panes(m).Pages(currentPageNumber).Breaks(x).range.
 -  Breaks do not correspond to columns.
- There is no range on a TextColumn object.

This is starting to seem quite difficult.

Rather than report the column number, would it be sufficient to announce the number of columns when it changes (eg in new sections)?

michaelDCurran commented:
range.information with wdHorizontalPositionRelativeToPage can give you distance from left edge of page in points.

feerrenrut committed:
Experiment to report the column number (#6559)

Now the number of text columns and the column number are reported in MS word.

This is done by getting the position relative to the start of the document (horizontally), and comparing it to the start position of the text column.
Since word does not have an interface to get the start position of the text column it is calculated. This is done by adding gutter, and margin, then the width of the column and space between to get each successive column.

feerrenrut commented:
@michaelDCurran I have updated this with the column count and current column number reporting. As discussed the column number is not reported when there is a gutter and mirror margins is set.

michaelDCurran commented on nvdaHelper/remote/WinWord/Constants.h:
```diff
@@ -128,38 +145,60 @@ constexpr int wdDISPID_TABLE_RANGE = 0;
 constexpr int wdInlineShapePicture = 3;
 constexpr int wdInlineShapeLinkedPicture = 4;
+
+// WdGutterStyle Enermeration
```
enermeration -> enumeration

michaelDCurran replied:
enermeration -> enumeration

michaelDCurran commented on nvdaHelper/remote/winword.cpp:
```diff
@@ -690,6 +694,272 @@ inline bool generateFootnoteEndnoteXML(IDispatch* pDispatchRange, wostringstream
 	return true;
 }
 
+std::experimental::optional<int> getPageBreakType(IDispatchPtr pDispatchRange ) {
```
Should this be perhaps called getSectionBreakType or some such? as from my reading it only returns something meaningful if the page break character particularly straddles 2 sections?

michaelDCurran commented on nvdaHelper/remote/winword.cpp:
```diff
@@ -690,6 +694,272 @@ inline bool generateFootnoteEndnoteXML(IDispatch* pDispatchRange, wostringstream
+}
+
+std::experimental::optional<int>
+getStartOfRangeDistanceFromStartOfDocument(IDispatchPtr pDispatchRange) {
```
the name "startOfDocument" bothers me, as  it could mean in terms of offsets. Perhaps "edgeOfDocument" may be better?

michaelDCurran commented:
Say you have a document with 2 sections. The first section has 1 column, and the second section has 2 columns. Arrowing down from the first section to the second section does not announce the fact that you are in a section with multiple columns. I wonder if it might be possible to  announce the number of columns along with the section number, rather than  with the column itself? For example, when arrowing down, it could say  "section 2, 2 columns, column 1". And then when changing columns just "column 2" (or nothing if it cannot calculate it).   I would also suggest dropping the "text" from "text column". Although its more specific,  it is adding a lot of extra speech which may slow people down.

feerrenrut commented:
It seems like we should add a tolerance to the width check to avoid spurious errors.

feerrenrut commented:
Also, note to self:

- [ ] update the user docs

feerrenrut commented:
Yep, thanks. I'll make those changes. Its intended that the number of columns is reported. 

- [ ] Test if column numbers are reported with "even page section" breaks.
- [ ] Are "even page section" breaks indicated visually to users somehow?

feerrenrut commented:
I couldn't reproduce the issue with "the even page section" breaks while in print view or in draft view. While in draft view there are visual indicators of section breaks, but not in print view.

feerrenrut committed:
Review actions for #6498

feerrenrut committed:
Reduced log level for debug message.

michaelDCurran commented:
I can no longer reproduce the issue where columns won't announce with even page section breaks. 
However, I think for reasons of verbosity, if a section only contains 1 column, then no column information should be exposed. So for example, when starting a new blank document in Microsoft Word, now the user will here page 1 section 1 column 1 of 1. This is a lot of extra info. Not exposing column info if there is only 1 column will not cause information loss as the column count can only change on a section boundary, and NVDA announces change sections.

michaelDCurran commented:
All the recent changes look good to me.

feerrenrut committed:
Be less verbose

When the number of columns has not changed, and there is just one column
do not report the column and column count.

feerrenrut commented:
I have modified this so that the column number and count is not reported when it has not changed and there is only one column

michaelDCurran commented:
Change looks good to me.

josephsl commented:
Hi @feerrenrut, I and others found a huge regression in Outlook 2016 (#6632):

STR:

1. In Outlook 2016, with latest next build, find an HTML message and a plain text message.
2. Open the HTML message and use say all.
3. Open plain text message and do say all.
Expected: no errors.
Actual: Traceback shown in #6632 is displayed and error tones are heard.

This also happens if one moves by line as well. Thanks.

feerrenrut committed:
Dont report columns in outlook

Fixed a regression that caused errors while using outlook. There is no
need to report page, section, or column numbers in outlook. We no longer
attempt to calculate them.

feerrenrut commented:
Thanks @josephsl. @michaelDCurran Could you take a look at the latest commit to fix this regression please?

feerrenrut commented on source/appModules/outlook.py:
```diff
@@ -175,7 +175,7 @@ def _get_shouldAllowIAccessibleFocusEvent(self):
 			return False
 		return super(SuperGridClient2010,self).shouldAllowIAccessibleFocusEvent
-
+
```
Yeah, my editor fixed some line endings.

josephsl commented:
Hi,

Next.13753/Outlook 2016 produces the following:

ERROR - RPC process 6212 (OUTLOOK.EXE) (18:41:52):
Thread 6216, build\x86_64\remote\winword.cpp, detectAndGenerateColumnFormatXML, 870:
error getting textColumn count. res: -2146823683 count: -1

This happens if one opens plain text messages. Thanks.

feerrenrut commented:
Hi @josephsl neither me or Mick can reproduce this, could you try rebooting your computer and test again?
If you are still getting this can you confirm the NVDA version and attach a log (with the level set to Input/Output) to this issue.

josephsl commented:
Hi, looks like the issue I mentioned (error tones in plain text messages) is gone now. Thanks.
